front/app/modules/commercial/insights/admin/containers/Insights/List/CreateInsightsView.tsx

```diff
@@ -53,41 +63,111 @@ interface InputProps {
   closeCreateModal: () => void;
 }
 
+const timeout = 400;
+
+const ArrowIcon = styled(Icon)`
+  flex: 0 0 12px;
+  width: 12px;
+  height: 12px;
+  transform: rotate(90deg);
+  transition: all 0.2s linear;
+  margin-left: 5px;
+
+  &.open {
+    transform: rotate(0deg);
+  }
+`;
+
+const AnimatedFieldset = styled.fieldset`
+  border: none;
+  margin: 0;
+  padding: 0;
+  opacity: 0;
+  display: none;
+  transition: all ${timeout}ms cubic-bezier(0.165, 0.84, 0.44, 1);
+  will-change: opacity, height;
+  border-bottom: 1px solid ${colors.separation};
+  margin-top: 15px;
+
+  &.collapse-enter {
+    opacity: 0;
+    max-height: 0px;
+    overflow: hidden;
+    display: block;
+
+    &.collapse-enter-active {
+      opacity: 1;
+      max-height: 350px;
+      overflow: hidden;
+      display: block;
+    }
+  }
+
+  &.collapse-enter-done {
+    opacity: 1;
+    overflow: visible;
+    display: block;
+  }
+
+  &.collapse-exit {
+    opacity: 1;
+    max-height: 350px;
+    overflow: hidden;
+    display: block;
+
+    &.collapse-exit-active {
+      opacity: 0;
+      max-height: 0px;
+      overflow: hidden;
+      display: block;
+    }
+  }
+
+  &.collapse-exit-done {
+    display: none;
+  }
+`;
+
+const StyledCheckboxWithPartialCheck = styled(CheckboxWithPartialCheck)`
+  label {
+    font-size: ${fontSizes.small}px;
+    color: ${colors.adminTextColor};
+  }
+`;
+
 export const CreateInsightsView = ({
   projects,
   closeCreateModal,
 }: DataProps & InputProps) => {
   const localize = useLocalize();
+  const { projectFolders } = useProjectFolders({});
   const [loading, setLoading] = useState(false);
   const [errors, setErrors] = useState<CLErrors | undefined>();
 
   const [name, setName] = useState<string | null>();
+
+  const [selectedProjectsIds, setSelectedProjectsIds] = useState<string[]>([]);
```

Comment:
Have you considered using `Set`s for these? Since you're using a lot of `selectedProjectsIds.include`, I figured maybe a `Set` might be a better data structure because `Set.has` has lookup time O(1) while `Array.include` has O(n). Also makes sense conceptually since all entries of `Set` are unique by definition, whereas `Array`s can contain duplicates. Of course that might add a bit more boilerplate in cases where you need to filter them, e.g. `[...selectedProjectsIds].filter` instead of `selectedProjectIds.filter`. Up to you!

Reply:
I did consider it and you are 100% right on the lookup time and the uniqueness of the items in the set. The issue is that I need a number of array methods - not only filter but some, every, adding multiple items at once - so to do any of this, I have to convert to array and back to set again which felt not optimal. Also, I could be wrong but I don't think the `includes` check will realistically result in performance issues. If it becomes a problem, we can always switch but for now, I would prefer the convenience of the array :)